### PR TITLE
CM - heat demand

### PR DESCRIPTION
**Pull request by cedricserugendo**

### CODE REVIEW

**Comment by neodark** [comment]
As stated on the branch:
`This branch is 29 commits ahead, 138 commits behind master.`

So you need to rebase your branch and push it back before we start reviewing your code.

File: cm/cm_heat_demand/tools/geofile.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    color_table.CreateColorRamp(
+        int(map_array.min()), (112, 153, 89), int(map_array.max()), (214, 193, 156)
+    )
+
+    band.SetRasterColorTable(color_table)
+    # band.SetRasterColorInterpretation(gdal.GCI_PaletteIndex)
+
+    driver = None
+    band = None
+    dst_ds = None
```
Comment (melhk) [inline cm/cm_heat_demand/tools/geofile.py:143]: What are you doing here?

File: cm/cm_heat_demand/tools/geofile.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    band.WriteArray(array=data_array_scaled)
+
+    # create the color table
+    color_table = gdal.ColorTable()
+    color_table.CreateColorRamp(
+        int(map_array.min()), (112, 153, 89), int(map_array.max()), (214, 193, 156)
+    )
+
+    band.SetRasterColorTable(color_table)
+    # band.SetRasterColorInterpretation(gdal.GCI_PaletteIndex)
```
Comment (melhk) [inline cm/cm_heat_demand/tools/geofile.py]: Remove commented lines please.

File: cm/cm_heat_demand/tools/response.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            base_dictionary["graphs"]["Areas potentials (GWh)"] = dict()
+            base_dictionary["graphs"]["Areas potentials (GWh)"]["type"] = "bar"
+
+            # labels start from 1, therefore the array size is 'num_labels_array + 1'
+            generator = [element for element in areas_potential[1:] if element > 0]
+
+            values = [
+                ("Zone " + str(index + 1), value)
+                for index, value in enumerate(generator)
+            ]
```
Comment (melhk) [inline cm/cm_heat_demand/tools/response.py:53]: I don't understand the comment. What are you doing here?

File: cm/cm_heat_demand/tools/response.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            * areas_path : path to the areas  geofile.
+
+        Output :
+            * base_dictionary : updated dictionary.
+        """
+
+        base_dictionary["geofiles"] = dict()
+        base_dictionary["geofiles"]["areas"] = areas_path
+
+        return base_dictionary
```
Comment (melhk) [inline cm/cm_heat_demand/tools/response.py]: What is this function used for? Are you printing a path in the CM output?

File: cm/cm_heat_demand/tools/areas.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    # array_map units : MWh
+
+    total_heat_demand = np.around(np.sum(array_map) / 1000, 2)
+    # total_heat_demand units : GWh
+
+    pixel_filtered_map = array_map * (array_map > pixel_threshold)
+    # pixel_filtered_map units : MWh
+
+    district_heating_zone_threshold = district_heating_zone_threshold * 1000
+    # district_heating_zone_threshold units : MWh (after the conversion)
```
Comment (melhk) [inline cm/cm_heat_demand/tools/areas.py]: Why are the comments below the code?

File: cm/cm_heat_demand/tools/geofile.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            raise Exception("Unable to load a raster")
+    else:
+        raise ValueError(f"This path does not refer to existing file : {raster}")
+
+    geo_transform = geofile.GetGeoTransform()
+    band = geofile.GetRasterBand(1)
+    array = band.ReadAsArray().astype(float)
+
+    geofile = None
+    band = None
```
Comment (melhk) [inline cm/cm_heat_demand/tools/geofile.py:88]: What are you doing here?

File: cm/cm_heat_demand/tools/geofile.py (hunk 1 above)
Comment (cedricserugendo) [reply cm/cm_heat_demand/tools/geofile.py:143]: GDAL expects us to close the dataset
https://gdal.org/api/python_gotchas.html#saving-and-closing-datasets-datasources

File: cm/cm_heat_demand/tools/geofile.py (hunk 2 above)
Comment (cedricserugendo) [reply cm/cm_heat_demand/tools/geofile.py]: it is done now

File: cm/cm_heat_demand/tools/response.py (hunk 3 above)
Comment (cedricserugendo) [reply cm/cm_heat_demand/tools/response.py:53]: I browse the list `areas_potential` containing the potential of each area by checking that the potential is positive 
Then I generate a tuple, see below
```
generator = [10, 15, ...]
values = [("Zone 1", 10), ("Zone 2", 15), ...]
```
The `values` is return to the frontend

File: cm/cm_heat_demand/tools/response.py (hunk 4 above)
Comment (cedricserugendo) [reply cm/cm_heat_demand/tools/response.py]: for the moment this function is empty for the moment but it should contain the address of the raster that the CM creates

File: cm/cm_heat_demand/tools/areas.py (hunk 5 above)
Comment (cedricserugendo) [reply cm/cm_heat_demand/tools/areas.py]: no particular reason, they are above the code now

File: cm/cm_heat_demand/tools/geofile.py (hunk 6 above)
Comment (cedricserugendo) [reply cm/cm_heat_demand/tools/geofile.py:88]: GDAL expects us to close the dataset
https://gdal.org/api/python_gotchas.html#saving-and-closing-datasets-datasources

File: cm/cm_heat_demand/tools/areas.py (hunk 5 above)
Comment (melhk) [reply cm/cm_heat_demand/tools/areas.py]: Thanks, it's more understandable to choose a way of doing things and stick to it!

File: cm/cm_heat_demand/tools/geofile.py (hunk 1 above)
Comment (melhk) [reply cm/cm_heat_demand/tools/geofile.py:143]: Ok

File: cm/cm_heat_demand/tools/response.py (hunk 3 above)
Comment (melhk) [reply cm/cm_heat_demand/tools/response.py:53]: But I don't understand the comment about the array size?

File: cm/cm_heat_demand/tools/response.py (hunk 4 above)
Comment (melhk) [reply cm/cm_heat_demand/tools/response.py]: What kind of address?

**Comment by melhk** [comment]
I get this error when trying to run the CM:

cm_not_working

File: cm/README.md [hunk 7]
```diff
@@ -68,38 +68,60 @@ Below is an example of the broker implementation.
...
+module has its own queue. 
+This queue is identified by a name given by the `get_default_app` function.
+
 The selected area (a geojson file) and the selected raster (a geotiff file)
 are in the variable ```selection``` and ```rasters``` respectively.
 
 See [test data](./multiply/testdata) for geojson file and geotiff file example.
 
 The variable ```params``` refers to the data provided by the form on the frontend.
+The method ```validate_params``` verify if the value enter by the user match with the schema
```
Comment (melhk) [inline cm/README.md]: valueS

File: cm/README.md [hunk 8]
```diff
@@ -68,38 +68,60 @@ Below is an example of the broker implementation.
...
 The selected area (a geojson file) and the selected raster (a geotiff file)
 are in the variable ```selection``` and ```rasters``` respectively.
 
 See [test data](./multiply/testdata) for geojson file and geotiff file example.
 
 The variable ```params``` refers to the data provided by the form on the frontend.
+The method ```validate_params``` verify if the value enter by the user match with the schema 
+of the ```schema.json``` file, see "Schema.json" section below.
 
 All CM provide a output dictionary. The 3 keys required are as follows:
```
Comment (melhk) [inline cm/README.md]: an + no uppercase in schema.json

File: cm/README.md [hunk 9]
```diff
@@ -68,38 +68,60 @@ Below is an example of the broker implementation.
...
 
 * graphs 
 * geofiles
 * values
 
+The schema of this dictionary have to be validated before it is sent to the frontend.
```
Comment (melhk) [inline cm/README.md]: has

File: cm/README.md [hunk 10]
```diff
@@ -68,38 +68,60 @@ Below is an example of the broker implementation.
...
 See [cm_output.py](./base/BaseCM/cm_output.py) for more information about the output json schema.
 
 Warning: the name of the function in the file `worker.py` corresponds to the name of the CM as it will appear in the front-end. Make sure that this name is unique, otherwise it will generate conflicts with other CMs.
```
Comment (melhk) [inline cm/README.md]: `worker.py` file

File: cm/cm_heat_demand/README.md [hunk 11]
```diff
@@ -0,0 +1,64 @@
+# References 
+This calculation module (CM) was originally written as part of the 
+[HotMaps](https://www.hotmaps-project.eu/) project.
+
+The documentation that was done for this first version can be found on
+(this section)[https://wiki.hotmaps.eu/en/CM-District-heating-potential-areas-user-defined-thresholds] 
+of the project HotMaps wiki.
```
Comment (melhk) [inline cm/cm_heat_demand/README.md]: Hotmaps project wiki

File: cm/cm_heat_demand/README.md [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+> work is in progress
+
+## Comparison with HotMaps plat
+> work is in progress
+
+# Example from front-end view
+> work is in progress
+
+## Set parameters (default value)
+> work is in progress
```
Comment (melhk) [inline cm/cm_heat_demand/README.md]: TODO

File: cm/cm_heat_demand/README.md [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+# Requirements 
+
+> Remark : 
+> GDAL is not in the requirements.txt file 
+> because install directly through command line in the Dockerfile
```
Comment (melhk) [inline cm/cm_heat_demand/README.md]: because it is installed

Comment (cedricserugendo) [reply cm/cm_heat_demand/README.md]: it's fixed

File: cm/cm_heat_demand/README.md (hunk 11 above)
Comment (cedricserugendo) [reply cm/cm_heat_demand/README.md]: it's fixed

File: cm/README.md (hunk 10 above)
Comment (cedricserugendo) [reply cm/README.md]: it's fixed

File: cm/README.md (hunk 9 above)
Comment (cedricserugendo) [reply cm/README.md]: it's fixed

File: cm/README.md (hunk 8 above)
Comment (cedricserugendo) [reply cm/README.md]: it's fixed

File: cm/README.md (hunk 7 above)
Comment (cedricserugendo) [reply cm/README.md]: it's fixed

Comment (melhk) [reply cm/README.md]: "verifies"

Comment (melhk) [reply cm/README.md]: "entered"

File: cm/README.md (hunk 8 above)
Comment (melhk) [reply cm/README.md]: I don't see that it is fixed?

File: cm/cm_heat_demand/README.md (hunk 12 above)
Comment (cedricserugendo) [reply cm/cm_heat_demand/README.md]: I have updated the file

File: cm/README.md (hunk 7 above)
Comment (cedricserugendo) [reply cm/README.md]: it's fixed

**Comment by neodark** [comment]
I deleted the containers, removed the volumes, cleaned up even all the images and went from scratch to make sure I have a clean setup and build up everything from your branch `cm_heat_demand`.

But I always get this:

![image](https://user-images.githubusercontent.com/2482166/130485634-cf12a5a5-d100-4ab4-bd76-ab7ea933f51a.png)

Looks like the CM is not in a working state.

Over that, some errors have appeared in the unit tests when building such as:

```
lling collected packages: wcwidth, prompt-toolkit, click-repl, vine, billiard, click-didyoumean, amqp, kombu, pytz, celery, pyrsistent, jsonschema, marshmallow, marshmallow-union, redis, BaseCM
Successfully installed BaseCM-0.0.1 amqp-5.0.6 billiard-3.6.4.0 celery-5.0.2 click-didyoumean-0.0.3 click-repl-0.2.0 jsonschema-3.2.0 kombu-5.1.0 marshmallow-3.9.1 marshmallow-union-0.1.15.post1 prompt-toolkit-3.0.20 pyrsistent-0.18.0 pytz-2021.1 redis-3.5.3 vine-5.0.0 wcwidth-0.2.5
......
----------------------------------------------------------------------                                                                                                                                            
Ran 6 tests in 0.359s                                                                                                                                                                                             

OK                                                                                                                                                                                                                
Removing intermediate container 5e9b6b28ec91
 ---> d853ad7acd99
Step 9/11 : COPY cm_heat_demand .
 ---> eb134534c75b
Step 10/11 : RUN python3 test.py
 ---> Running in c32e1c9464be
......F.
======================================================================                                                                                                                                            
FAIL: test_write_raster (tools.test_geofile.TestGeofileTools)                                                                                                                                                     
Copy a raster and read the copy as well as the original to make sure that                                                                                                                                         
----------------------------------------------------------------------                                                                                                                                            
Traceback (most recent call last):                                                                                                                                                                                
  File "/cm-heat-demand/tools/test_geofile.py", line 96, in test_write_raster                                                                                                                                     
    self.comparator(map_array, dst_map_array, comparison=self.array_comparison)                                                                                                                                   
  File "/cm-heat-demand/tools/test_geofile.py", line 44, in comparator                                                                                                                                            
    self.assertTrue(False)                                                                                                                                                                                        
AssertionError: False is not true                                                                                                                                                                                 

----------------------------------------------------------------------                                                                                                                                            
Ran 8 tests in 0.127s                                                                                                                                                                                             

FAILED (failures=1)                                                                                                                                                                                               
Removing intermediate container c32e1c9464be
 ---> 052505298402
Step 11/11 : CMD ["python3", "worker.py"]
 ---> Running in bf0e9289d718
Removing intermediate container bf0e9289d718
 ---> 16ba788edbfc
Successfully built 16ba788edbfc
Successfully tagged enermaps_cm-heat-demand:latest
Building cm-multiply
Sending build context to Docker daemon  10.48MB
Step 
```

So there's some issue there too to check.

**Comment by cedricserugendo** [comment]
@neodark
The CM takes a raster as input. I think you have a vector.
To test the CM, I suggest to use dataset 43 which corresponds to the layer expected by the CM.

I can't manage to recreate the error in the unit tests.
@mel, do you have the same problems with these tests?

**Comment by melhk** [comment]
I still have the same error, even when I select layer 43

Capture
 layer 43:

**Comment by PuckCh** [comment]
> I still have the same error, even when I select layer 43
> 
> Capture
> layer 43:

I had the same problem, and it was due to space in raster filenames.
I have commited a patch to fix the problem (I hope).

**Comment by PuckCh** [comment]
Selection by country is not working. Error on the interface is:
`local variable 'resp' referenced before assignment`

The error with the variable is on file cm/base/BaseCM/cm_output.py, in function output_raster.
When the exception ConnectionError raise, resp is not defined.
You need to:
1. correct the case with the ConnectionError occur,
2. see why this exception come for country selection.

File: cm/cm_heat_demand/README.md [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+### Demand in DH
+The studied layer contains one or more areas where the potential is null.
+The CM will then apply successively a binary dilatation 
+(see [scipy page](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.ndimage.morphology.binary_dilation.html) 
+for more information) and erosion operation
+(see [scipy page](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.ndimage.morphology.binary_erosion.html) 
+for more information) to obtain the final layer.
+
+Then, the CM sum up  zone by zone the value of the remaining pixels .
```
Comment (melhk) [inline cm/cm_heat_demand/README.md]: will sum / the values

File: cm/cm_heat_demand/README.md [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+If the sum of a zone is less than a given threshold, 
+the pixel values defining this zone are being  to zero.
```
Comment (melhk) [inline cm/cm_heat_demand/README.md]: word missing?

File: cm/cm_heat_demand/test.py [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+        return unittest.TestLoader().loadTestsFromTestCase(tests_class)
+
+    test_case = [import_testcase(testcase) for testcase in testcases]
+
+    test_suite = unittest.TestSuite(test_case)
+
+    return test_suite
+
+
+testsuite = get_testsuite(TestSettings, TestGeofileTools, TestAreasTools)
```
Comment (melhk) [inline cm/cm_heat_demand/test.py]: This variable must be in the main function.

File: cm/cm_heat_demand/tools/areas.py [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+    structure = np.ones((3, 3)).astype(int)
+    expanded_map = binary_dilation(input=pixel_filtered_map, structure=structure)
+    eroded_map = binary_erosion(input=expanded_map, structure=structure)
+    labels_array, n_label = measurements.label(
+        input=eroded_map,
+        structure=structure,
+    )
+
+    # labels start from 1, therefore the array size is 'num_labels_array + 1'
+    areas_potential = np.zeros((n_label + 1)).astype(float)
```
Comment (melhk) [inline cm/cm_heat_demand/tools/areas.py:105]: I still don't understand this comment about the array size. Please explain me what it means.

File: cm/cm_heat_demand/tools/areas.py [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+    if n_label > 0:
+        end, start, sorted_array = get_browsing_indexes(
+            labels_array=labels_array,
+            pixel_filtered_map=pixel_filtered_map,
+            n_label=n_label,
+        )
+
+        for i, (start_, end_) in enumerate(zip(start, end)):
```
Comment (melhk) [inline cm/cm_heat_demand/tools/areas.py]: What is the meaning of this _ at the end of the variables?

File: cm/cm_heat_demand/tools/geofile.py [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+
+import numpy as np
+import rasterio
+from osgeo import gdal, osr
+
+
+def get_projection(geofile: str):
+    """Get the projection of a geofile."""
+
+    with rasterio.open(geofile) as src_file:
```
Comment (melhk) [inline cm/cm_heat_demand/tools/geofile.py:15]: Check that you have the expected file type? That the file opening succeeds?

File: cm/cm_heat_demand/tools/geofile.py [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+    band = geofile.GetRasterBand(1)
+    array = band.ReadAsArray().astype(float)
+
+    geofile = None
+    band = None
+
+    if return_geo_transform:
+        return array, geo_transform
+    else:
+        return array
```
Comment (melhk) [inline cm/cm_heat_demand/tools/geofile.py:93]: Why not always return the "geo transform"?

File: cm/cm_heat_demand/tools/response.py [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+
+        # Areas potential
+        if areas_potential[1:].size > 0:
+            print("areas_potential")
+            print(areas_potential[1:])
+            base_dictionary["graphs"]["Areas potentials (GWh)"] = dict()
+            base_dictionary["graphs"]["Areas potentials (GWh)"]["type"] = "bar"
+
+            # labels start from 1, therefore the array size is 'num_labels_array + 1'
+            generator = [element for element in areas_potential[1:] if element > 0]
```
Comment (melhk) [inline cm/cm_heat_demand/tools/response.py:48]: What does this comment mean?

File: cm/cm_heat_demand/worker.py [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+    if not rasters:
+        raise ValueError("Raster list must be non-empty.")
+    if "features" not in selection:
+        raise ValueError("The selection must be a feature set.")
+    if not selection["features"]:
+        raise ValueError("The selection must be non-empty.")
+    self.validate_params(params)
+
+    raster = cm_input.get_raster_path(rasters[0])
+    region = selection["features"][0]["geometry"]
```
Comment (melhk) [inline cm/cm_heat_demand/worker.py]: Check the input parameters?

**Comment by melhk** [comment]
> o fix the problem

The patch of @PuckCh works for me too, but I get an error when I try to display the layer that the cm has created. @cedricserugendo, you should check that the raster that is created is a valid raster (maybe try to copy it locally and open it).

File: cm/cm_heat_demand/tools/geofile.py (hunk 20 above)
Comment (cedricserugendo) [reply cm/cm_heat_demand/tools/geofile.py:93]: we don't always need it

File: cm/cm_heat_demand/tools/areas.py (hunk 18 above)
Comment (cedricserugendo) [reply cm/cm_heat_demand/tools/areas.py]: I didn't want to have two variables with the same name, I modified this

File: cm/cm_heat_demand/README.md (hunk 15 above)
Comment (cedricserugendo) [reply cm/cm_heat_demand/README.md]: I changed the wording

File: cm/cm_heat_demand/README.md (hunk 14 above)
Comment (cedricserugendo) [reply cm/cm_heat_demand/README.md]: it is fixed

File: cm/cm_heat_demand/tools/response.py (hunk 4 above)
Comment (cedricserugendo) [reply cm/cm_heat_demand/tools/response.py]: for example: /api/geofile/{laye name}
I corrected this

Comment (cedricserugendo) [reply cm/cm_heat_demand/tools/response.py]: for example: /api/geofile/{filename}

File: cm/cm_heat_demand/test.py (hunk 16 above)
Comment (cedricserugendo) [reply cm/cm_heat_demand/test.py]: it's done

File: cm/cm_heat_demand/tools/geofile.py (hunk 19 above)
Comment (cedricserugendo) [reply cm/cm_heat_demand/tools/geofile.py:15]: it's done now

File: cm/cm_heat_demand/worker.py (hunk 22 above)
Comment (cedricserugendo) [reply cm/cm_heat_demand/worker.py]: it's done now

**Comment by cedricserugendo** [comment]
Apparently, the error was caused by spaces in the raster names
Thanks @PuckCh for finding the solution to this problem

I've made a commit to handle the 'ConnectionError'
Concerning the country selection, the problem comes from the country size 
If we select a smaller country, the CM works

As for the unit tests, I had forgotten a `self.assertTrue(False)`, it has been removed
Now, @neodark you should not see this error anymore

**Comment by cedricserugendo** [comment]
@melhk here is the explanation for the size of `areas_potential` variable in the function `define_areas` (/tools/areas.py)


In the function `define_areas` we define a matrix in which each value corresponds to the index of the area to which the pixel belongs. This is the `labels_array` variable. Then we obtain the number of labels found. This is the `n_label` variable.

Then, after different manipulations, we've got the variable `areas_potential` which is in fact a list whose first element is equal to zero. The other elements of this list correspond to the potential of the areas.

The line `areas = areas_potential[labels_array]` will assign the value 0 to all the elements of `labels_array` equal to 0, then the value of the first area to the elements having the value 1 (thus belonging to the first area), etc.

Here is some examples :
```python
areas_potential = array([ 0., 10., 20., 30., 40.])
labeled_array = array([[0, 0, 1, 1, 0, 0],
                       [0, 0, 0, 1, 0, 0],
                       [2, 2, 0, 0, 3, 0],
                       [0, 0, 0, 4, 0, 0]])
areas_potential[labeled_array] = array([[ 0.,  0., 10., 10.,  0.,  0.],
                                        [ 0.,  0.,  0., 10.,  0.,  0.],
                                        [20., 20.,  0.,  0., 30.,  0.],
                                        [ 0.,  0.,  0., 40.,  0.,  0.]])
```

```python
areas_potential = array([ 999., 10., 20., 30., 40.])
labeled_array = array([[0, 0, 1, 1, 0, 0],
                       [0, 0, 0, 1, 0, 0],
                       [2, 2, 0, 0, 3, 0],
                       [0, 0, 0, 4, 0, 0]])
areas_potential[labeled_array] = array([[999., 999.,  10.,  10., 999., 999.],
                                        [999., 999., 999.,  10., 999., 999.],
                                        [ 20.,  20., 999., 999.,  30., 999.],
                                        [999., 999., 999.,  40., 999., 999.]])
```

File: cm/cm_heat_demand/tools/response.py (hunk 3 above)
Comment (cedricserugendo) [reply cm/cm_heat_demand/tools/response.py:53]: I answered in this comment: https://github.com/enermaps/enermaps/pull/256#issuecomment-908196761

File: cm/cm_heat_demand/tools/areas.py (hunk 17 above)
Comment (cedricserugendo) [reply cm/cm_heat_demand/tools/areas.py:105]: I answered in this comment: https://github.com/enermaps/enermaps/pull/256#issuecomment-908196761

File: cm/cm_heat_demand/tools/response.py (hunk 21 above)
Comment (cedricserugendo) [reply cm/cm_heat_demand/tools/response.py:48]: I answered in this comment: https://github.com/enermaps/enermaps/pull/256#issuecomment-908196761

**Review by melhk — APPROVED** [review]

**Pull request merged**